cpp/include/raft/buffer_base.hpp

```diff
@@ -0,0 +1,183 @@
+/*
+ * Copyright (c) 2019-2020, NVIDIA CORPORATION.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#pragma once
+
+#include <cuda_runtime.h>
+#include <memory>
+#include "cudart_utils.h"
+
+namespace raft {
+
+/**
+ * @brief Base for all RAII-based owning of temporary memory allocations. This
+ *        class should ideally not be used by users directly, but instead via
+ *        the child classes `device_buffer` and `host_buffer`.
+ */
+template <typename T, typename Allocator>
+class buffer_base {
```

Comment:
And why do we have this considering that rmm has device_buffer?

Reply:
rmm certainly has a device_buffer, but it doesn't have `host_buffer` (afaik). We need both for best performance.

Reply:
also, using rmm's device_buffer means we are tied to only rmm, but this interface is generic enough to support any allocator.

Reply:
rmm supports multiple different allocators, and I was worried about reinventing the wheel and duplicating efforts; and I thought RAPIDS projects are tied to rmm, but this is just my 2 cents :-)

Reply:
cuML is NOT tied to rmm :) For eg: one can even use the basic cub caching-allocator too for all allocations done inside cuML and it'll work just fine!

Do you know if rmm supports host-memory (cudaMallocHost) based allocators? AFAIK, this is not the case.

Reply:
> Do you know if rmm supports host-memory (cudaMallocHost) based allocators?

Yeah... AFAIK, rmm currently does not support pinned memory allocation on host.

Reply:
My biggest concern is that RAPIDS may need to develop support for new features (e.g. something like https://github.com/rapidsai/raft/issues/6) in both rmm and raft as such features will be helpful to both rmm & raft users.

Reply:
It does appear that RMM does not support host_allocator in the way we want it to be, ie. to have support for streams in its interface. So, we are better off for now to have this in raft and later work with rmm to update this interface.